src/databricks/labs/ucx/cli.py

```diff
@@ -292,19 +292,25 @@ def create_uber_principal(
     w: WorkspaceClient,
     prompts: Prompts,
     ctx: WorkspaceContext | None = None,
+    run_as_collection: bool = False,
+    a: AccountClient | None = None,
     **named_parameters,
 ):
     """For azure cloud, creates a service principal and gives STORAGE BLOB READER access on all the storage account
     used by tables in the workspace and stores the spn info in the UCX cluster policy. For aws,
     it identifies all s3 buckets used by the Instance Profiles configured in the workspace.
     Pass subscription_id for azure and aws_profile for aws."""
-    if not ctx:
-        ctx = WorkspaceContext(w, named_parameters)
-    if ctx.is_azure:
-        return ctx.azure_resource_permissions.create_uber_principal(prompts)
-    if ctx.is_aws:
-        return ctx.aws_resource_permissions.create_uber_principal(prompts)
-    raise ValueError("Unsupported cloud provider")
+    if ctx:
+        workspace_contexts = [ctx]
+    else:
+        workspace_contexts = _get_workspace_contexts(w, a, run_as_collection, **named_parameters)
+    for workspace_context in workspace_contexts:
+        if workspace_context.is_azure:
+            workspace_context.azure_resource_permissions.create_uber_principal(prompts)
```

Comment:
@nfx : For the uber principal, I think we should (consider to) create one uber principal for all workspaces, instead of one per workspace. The word "uber" could (also) suggest "over" all workspaces (and storage accounts).

It means moving the for-loop over workspace contexts into the `create_uber_principal` invocation. However, it would require as that is part of the workspace context, which contains one workspace context by definition. A refactor is required to achieve this. (Linking #2507 to log this)

Reply:
Taking this concern out of the scope of this PR: https://github.com/databrickslabs/ucx/issues/2644